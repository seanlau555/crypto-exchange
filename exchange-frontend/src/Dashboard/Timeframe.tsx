import { Flex, Button } from '@chakra-ui/react'
import { ButtonEvent } from '../types'

const availableTimeframe = ['1Min', '15Min', '1Hour', '1Day', '1Week', '1Month']

type Props = {
  onChange: (value: string) => void
  currentTimeframe: string
}

function Timeframe({ onChange, currentTimeframe }: Props) {
  const onClick = (timeframe: string, evt: ButtonEvent) => {
    evt.preventDefault()
    onChange(timeframe)
  }

  return (
    <Flex>
      {availableTimeframe.map((x) => (
        <Button
          key={x}
          colorScheme={currentTimeframe === x ? 'teal' : 'gray'}
          onClick={(evt) => onClick(x, evt)}
        >
          {x}
        </Button>
      ))}
    </Flex>
  )
}

export default Timeframe
