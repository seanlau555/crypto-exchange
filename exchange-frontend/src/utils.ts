import axios, { AxiosResponse } from 'axios'
import { Bar } from './types'

const Utils = {
  async getAuthToken(oauth_code: string) {
    // returns Authorization Token once we have our OAuth token
    const body = {
      grant_type: 'authorization_code',
      code: oauth_code,
      client_id: import.meta.env.VITE_REACT_APP_CLIENT_ID,
      client_secret: import.meta.env.VITE_REACT_APP_CLIENT_SECRET,
      redirect_uri: import.meta.env.VITE_REACT_APP_REDIRECT_URI,
    }
    // encode data into form encoding
    const encodedBody = Object.keys(body)
      .map((key: string) => `${key}=${encodeURIComponent(body[key])}`)
      .join('&')
    console.log(body)
    // submit POST request
    const response = await axios({
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      method: 'POST',
      url: 'https://api.alpaca.markets/oauth/token',
      data: encodedBody,
    })

    const { data } = response
    window.localStorage.setItem('auth-token', data)
    return data.access_token
  },

  parseResponse(response: AxiosResponse): Array<Bar> | [] {
    const { bars } = response.data
    if (bars) {
      const data = []
      for (let i = 0; i < bars.length; i++) {
        const bar = bars[i]
        const point = {
          date: bar.t,
          open: bar.o,
          low: bar.l,
          high: bar.h,
          close: bar.c,
          volume: bar.v,
        }
        data.push(point)
      }
      return data
    }
    return []
  },
}
export default Utils
