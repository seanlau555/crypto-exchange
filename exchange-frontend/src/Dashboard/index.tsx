import { useState } from 'react'
import CandleStickChart from './Chart'
import Timeframe from './Timeframe'
import Utils from '../Utils'
import { Flex, Button, Input } from '@chakra-ui/react'
import Sidebar from './Sidebar'
import styled from '@emotion/styled'
import { InputEvent } from '../types'
import { useGetCurrencyBars, useGetCurrencySubscription } from '../services'

function Dashboard() {
  const [inputValue, setInputValue] = useState<string>('')
  const [symbol, setSymbol] = useState<string>('BTCUSD')
  const [timeframe, setTimeframe] = useState<string>('1Day')
  const token = window.localStorage.getItem('auth-token') || ''
  const { data } = useGetCurrencyBars(symbol, timeframe, token)

  useGetCurrencySubscription(symbol, token)

  const handleChange = (evt: InputEvent) => {
    setInputValue(evt.target.value)
  }

  const handleSubmit = async (e) => {
    e.preventDefault()

    if (window.localStorage.getItem('auth-token') === null) {
      var oauth_code = new URLSearchParams(window.location.search).get('code')
      const auth_token = await Utils.getAuthToken(oauth_code)
      window.localStorage.setItem('auth-token', auth_token)
    }
    setSymbol(inputValue)
  }

  const onChangeTimeframe = (value: string) => {
    setTimeframe(value)
  }

  return (
    <Flex bg="#fef7d0" height="100vh">
      <Sidebar selectedTicker={symbol} onSelect={setSymbol} />
      <div>
        <Flex>
          <StyledForm onSubmit={handleSubmit}>
            <label>
              Symbol:
              <Input
                bg="white"
                textTransform="uppercase"
                type="text"
                onChange={handleChange}
              />
            </label>
            <Button type="submit" colorScheme="teal" ml="8px">
              Submit
            </Button>
          </StyledForm>
          <Timeframe
            currentTimeframe={timeframe}
            onChange={onChangeTimeframe}
          />
        </Flex>
        <div className="chart">
          <label className="chart-symbol">
            Current Symbol: <b> {symbol.toUpperCase()} </b>{' '}
          </label>
          <CandleStickChart data={data} />
        </div>
      </div>
    </Flex>
  )
}

export default Dashboard

const StyledForm = styled.form`
  display: flex;
  width: 400px;
  label {
    display: flex;
    align-items: center;
    input {
      margin-left: 8px;
    }
  }
`
