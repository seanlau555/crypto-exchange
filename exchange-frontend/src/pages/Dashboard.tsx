import { useEffect, useState } from 'react'
import CandleStickChart from '../components/Chart'
import Timeframe from '../components/Timeframe'
import Utils from '../utils'
import { Flex } from '@chakra-ui/react'
import Sidebar from '../components/Sidebar'
import InputBar from '../components/InputBar'
import { useGetCurrencyBars } from '../services'
import { useNavigate } from 'react-router-dom'

const defaultTimeframe = '1Day'

function Dashboard() {
  const [symbol, setSymbol] = useState<string>('')
  const [timeframe, setTimeframe] = useState<string>(defaultTimeframe)
  const token = window.localStorage.getItem('auth-token') || ''
  const { data, refetch } = useGetCurrencyBars(symbol, timeframe, token)
  const navigate = useNavigate()

  useEffect(() => {
    async function getToken() {
      if (window.localStorage.getItem('auth-token') === null) {
        const oauth_code = new URLSearchParams(window.location.search).get(
          'code',
        )
        if (oauth_code) {
          const auth_token = await Utils.getAuthToken(oauth_code)
          window.localStorage.setItem('auth-token', auth_token)
          refetch()
        } else {
          navigate('/')
        }
      }
    }
    getToken()
  }, [])

  const handleSubmit = (value: string) => {
    setSymbol(value)
  }

  const onChangeTimeframe = (value: string) => {
    setTimeframe(value)
  }

  return (
    <Flex height="100vh">
      <Sidebar selectedTicker={symbol} onSelect={setSymbol} />
      <div>
        <Flex>
          <InputBar onSubmit={handleSubmit} />
          <Timeframe
            defaultTimeframe={defaultTimeframe}
            onChange={onChangeTimeframe}
          />
        </Flex>
        {data && (
          <div className="chart">
            <label className="chart-symbol">
              Current Symbol: <b> {symbol.toUpperCase()} </b>{' '}
            </label>
            <CandleStickChart inputData={data} />
          </div>
        )}
      </div>
    </Flex>
  )
}

export default Dashboard
