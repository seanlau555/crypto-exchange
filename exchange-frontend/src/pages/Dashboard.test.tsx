import { describe, expect, it } from 'vitest'
import Dashboard from './Dashboard'
// import { render, screen } from '@testing-library/react'
// import userEvent from '@testing-library/user-event'
import { createWrapper } from '../test/utils'

describe('<Dashboard />', () => {
  it('render dashbard page', () => {
    const { container } = createWrapper(<Dashboard />)
    expect(container).toMatchSnapshot()
  })
})
