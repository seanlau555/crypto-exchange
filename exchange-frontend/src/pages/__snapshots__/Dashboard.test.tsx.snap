// Vitest Snapshot v1

exports[`<Dashboard /> > render dashbard page 1`] = `
<div>
  <div
    class="css-1yeiifd"
  >
    <div
      class="css-1hn0ibo"
    >
      <div
        class="css-1j9jo68"
        id="BTCUSD"
      >
        BTCUSD
        , 
        Price: USD undefined
      </div>
      <div
        class="css-bhtpg9"
        id="ETHUSD"
      >
        ETHUSD
        , 
        Price: USD undefined
      </div>
      <div
        class="css-bhtpg9"
        id="SOLUSD"
      >
        SOLUSD
        , 
        Price: USD undefined
      </div>
      <div
        class="css-bhtpg9"
        id="DOGEUSD"
      >
        DOGEUSD
        , 
        Price: USD undefined
      </div>
      <div
        class="css-bhtpg9"
        id="AVAXUSD"
      >
        AVAXUSD
        , 
        Price: USD undefined
      </div>
      <div
        class="css-bhtpg9"
        id="SANDUSD"
      >
        SANDUSD
        , 
        Price: USD undefined
      </div>
    </div>
    <div>
      <div
        class="css-k008qs"
      >
        <form
          class="css-1qhrdhp"
        >
          <label>
            Symbol:
            <input
              class="chakra-input css-fmnpk6"
              type="text"
            />
          </label>
          <button
            class="chakra-button css-13fg2ty"
            type="submit"
          >
            Submit
          </button>
        </form>
        <div
          class="css-k008qs"
        >
          <button
            class="chakra-button css-5c2176"
            type="button"
          >
            1Min
          </button>
          <button
            class="chakra-button css-5c2176"
            type="button"
          >
            15Min
          </button>
          <button
            class="chakra-button css-5c2176"
            type="button"
          >
            1Hour
          </button>
          <button
            class="chakra-button css-5c2176"
            type="button"
          >
            1Day
          </button>
          <button
            class="chakra-button css-5c2176"
            type="button"
          >
            1Week
          </button>
          <button
            class="chakra-button css-5c2176"
            type="button"
          >
            1Month
          </button>
        </div>
      </div>
      <div
        class="chart"
      >
        <label
          class="chart-symbol"
        >
          Current Symbol: 
          <b>
             
            BTCUSD
             
          </b>
           
        </label>
        <div />
      </div>
    </div>
  </div>
</div>
`;
